### PR TITLE
Implement Tasks tab in the woman's profile page

### PR DESCRIPTION
**Pull request by dubdabasoduba**

### CODE REVIEW

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ContactPresenter.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755&bid=16503220)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ContactPresenter.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755&bid=16503220)

File: opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java [hunk 1]
```diff
@@ -84,24 +92,87 @@ protected void startFormActivity(JSONObject form, Contact contact) {
...
 
-        intent.putExtra(ConstantsUtils.JsonFormExtraUtils.JSON, getFormJson(partialContactRequest, form));
-        intent.putExtra(JsonFormConstants.JSON_FORM_KEY.FORM, contact);
-        intent.putExtra(ConstantsUtils.IntentKeyUtils.BASE_ENTITY_ID,
-                getIntent().getStringExtra(ConstantsUtils.IntentKeyUtils.BASE_ENTITY_ID));
-        intent.putExtra(ConstantsUtils.IntentKeyUtils.CLIENT_MAP, getIntent().getSerializableExtra(ConstantsUtils.IntentKeyUtils.CLIENT_MAP));
-        intent.putExtra(ConstantsUtils.IntentKeyUtils.FORM_NAME, contact.getFormName());
-        intent.putExtra(ConstantsUtils.IntentKeyUtils.CONTACT_NO, contactNo);
-        startActivityForResult(intent, JsonFormUtils.REQUEST_CODE_GET_JSON);
+    private String getUpdatedTestForm(JSONObject form, Contact contact, PartialContact partialContactRequest) throws JSONException {
```
Comment (keymane) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java]: Codacy Issue found: [Avoid reassigning parameters such as 'form'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ContactPresenter.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755&bid=16503220)

File: opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java (hunk 1 above)
Comment (keymane) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java]: Codacy Issue found: [Avoid reassigning parameters such as 'form'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java  8
- opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ContactPresenter.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755&bid=16503220)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ContactPresenter.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755&bid=16503220)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java  8
- opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ContactPresenter.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755&bid=16503220)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java  8
- opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ContactPresenter.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileTasksFragment.java  1
         

Clones added
============
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java  8
- opensrp-anc/src/test/java/org/smartregister/anc/library/fragment/ProfileTasksFragmentTest.java  12
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755&bid=16503220)

File: opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+import org.robolectric.android.controller.ActivityController;
+import org.smartregister.anc.library.AncLibrary;
+import org.smartregister.anc.library.model.PartialContact;
+import org.smartregister.anc.library.repository.PartialContactRepositoryHelper;
+import org.smartregister.anc.library.util.ConstantsUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
```
Comment (keymane) [inline opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java]: Codacy Issue found: [Avoid unused imports such as 'java.util.List'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

File: opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+import org.powermock.reflect.Whitebox;
+import org.robolectric.Robolectric;
+import org.robolectric.RuntimeEnvironment;
+import org.robolectric.android.controller.ActivityController;
+import org.smartregister.anc.library.AncLibrary;
+import org.smartregister.anc.library.model.PartialContact;
+import org.smartregister.anc.library.repository.PartialContactRepositoryHelper;
+import org.smartregister.anc.library.util.ConstantsUtils;
+
+import java.util.ArrayList;
```
Comment (keymane) [inline opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java]: Codacy Issue found: [Avoid unused imports such as 'java.util.ArrayList'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

File: opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
+package org.smartregister.anc.library.activity;
+
+import android.app.Activity;
+import android.content.Intent;
+import android.os.Bundle;
```
Comment (keymane) [inline opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java]: Codacy Issue found: [Avoid unused imports such as 'android.os.Bundle'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

File: opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+import java.util.Set;
+
+import timber.log.Timber;
+
+public class MainContactActivityTest extends BaseActivityUnitTest {
+    private ActivityController<MainContactActivity> activityController;
+    private MainContactActivity activity;
+
+    @Mock
+    private PartialContactRepositoryHelper partialContactRepositoryHelper;
```
Comment (keymane) [inline opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java]: Codacy Issue found: [Avoid unused private fields such as 'partialContactRepositoryHelper'.](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

File: opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+import android.content.Intent;
+import android.os.Bundle;
+import android.widget.TextView;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.ArgumentMatchers;
```
Comment (keymane) [inline opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.mockito.ArgumentMatchers'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

File: opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.powermock.reflect.Whitebox;
+import org.robolectric.Robolectric;
+import org.robolectric.RuntimeEnvironment;
+import org.robolectric.android.controller.ActivityController;
+import org.smartregister.anc.library.AncLibrary;
+import org.smartregister.anc.library.model.PartialContact;
```
Comment (keymane) [inline opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.smartregister.anc.library.model.PartialContact'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

File: opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+public class MainContactActivityTest extends BaseActivityUnitTest {
+    private ActivityController<MainContactActivity> activityController;
+    private MainContactActivity activity;
+
+    @Mock
+    private PartialContactRepositoryHelper partialContactRepositoryHelper;
+
+    @Mock
+    private AncLibrary instance;
```
Comment (keymane) [inline opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java]: Codacy Issue found: [Avoid unused private fields such as 'instance'.](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ContactPresenter.java  1
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java  8
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileTasksFragment.java  1
         

Clones added
============
- opensrp-anc/src/test/java/org/smartregister/anc/library/fragment/ProfileTasksFragmentTest.java  12
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755&bid=16503220)

File: opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.powermock.reflect.Whitebox;
+import org.robolectric.Robolectric;
+import org.robolectric.RuntimeEnvironment;
+import org.robolectric.android.controller.ActivityController;
+import org.smartregister.anc.library.AncLibrary;
+import org.smartregister.anc.library.repository.PartialContactRepositoryHelper;
```
Comment (keymane) [inline opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.smartregister.anc.library.repository.PartialContactRepositoryHelper'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

File: opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+import android.app.Activity;
+import android.content.Intent;
+import android.widget.TextView;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
```
Comment (keymane) [inline opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.mockito.Mock'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

File: opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+import org.mockito.Mockito;
+import org.powermock.reflect.Whitebox;
+import org.robolectric.Robolectric;
+import org.robolectric.RuntimeEnvironment;
+import org.robolectric.android.controller.ActivityController;
+import org.smartregister.anc.library.AncLibrary;
```
Comment (keymane) [inline opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.smartregister.anc.library.AncLibrary'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileTasksFragment.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java  8
- opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ContactPresenter.java  1
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java  3
         

Clones added
============
- opensrp-anc/src/test/java/org/smartregister/anc/library/fragment/ProfileTasksFragmentTest.java  12
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755&bid=16503220)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/BaseContactActivity.java  8
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileTasksFragment.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ContactPresenter.java  1
         

Clones added
============
- opensrp-anc/src/test/java/org/smartregister/anc/library/fragment/ProfileTasksFragmentTest.java  12
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/MainContactActivityTest.java  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5001755&bid=16503220)

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/JsonFormUtils.java [hunk 12]
```diff
@@ -829,6 +829,37 @@ public static JSONObject readJsonFromAsset(Context context, String filePath) thr
...
     }
 
+    /**
+     * Gets an expansion panel {@link JSONObject} value then check whether the test/tasks was `ordered` or `not done`.
+     * If any either of this is selected then we mark it as not complete.
+     *
+     * @param field {@link JSONObject}
+     * @return isComplete {@link Boolean}
+     */
+    public static boolean checkIfTaskIsComplete(JSONObject field) {
```
Comment (ellykits) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/util/JsonFormUtils.java]: You might later consider using `EasyRules`

**Pull request merged**